pkg/database/extensions_test.go

```diff
@@ -0,0 +1,94 @@
+package database
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"reflect"
+	"testing"
+
+	"github.com/ugorji/go/codec"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/encrypt"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+	"github.com/Azure/ARO-RP/test/util/manifests"
+)
+
+func TestUnmarshalSecure(t *testing.T) {
```

Comment:
testing needs more attention.  Possibly the strategy might need to be:
* use a real Database object in test
* make `h` a real field in the Database object
* make rand.Reader pluggable and use a dummy implementation which returns zeros
* use a fixed key value
* use Database.h to test serialisations and deserialisations to/from known values

Possibly the known values bit belongs in pkg/util/encryption.  Perhaps in the database layer you will test with a mock Cipher.

See what ideas @m1kola has about this.

Reply:
Not sure of the value of the additional complexity here. Making `h` part of the Database structure will not solve the suggested test path. `h` in the database layer is used deeper in the code and is coming from `NewDatabaseClient` method params. This way mocking `DatabaseClient` is not an option as we would be abstracting it away from my mock. Mocking it would make sense if we test backend, and we need to hook-in database calls into it. I think we want to test Document transition here. And this is what this test is doing. We can extend it to make sure intermediate value is as expected too. If we want better testing we should add this as a requirement when we write backend calls, as at that level we can use read `Database` object, and mocked individual clients. Which should give us wanted outcomes.

Ok, some thinking, while trying this:
* Added `h` to the Database struct. But we can't call `NewDatabase` in the tests because individual client creation is calling triggers API to create a client. And mocking cant help here due to the structure of the Dabatase client.
* Rand.Reader pluggable (I assume you have in mind rand.Rand one) does not make sense, as even with `zero` key and nonce result is always different. So testing `expected` results do not work.
* I added a new method to `database.NewJSONHandle(cipher)` and unified around the code, so we configure it only in once place.

Current tests in Encryption package tests `TestEncryptRoundTrip` and `TestExtensions` covers most of the cases. If we want more, can we do it as separate PR as this is already big?

Reply:
> Rand.Reader pluggable (I assume you have in mind rand.Rand one)

No, in test this would have to return something constant - like all zeros.